modepy/shapes.py

```diff
@@ -559,4 +566,40 @@ def _submes_for_hypercube(shape: Hypercube, node_tuples):
 # }}}
 
 
+# {{{ tensor product shape
+
+@dataclass(frozen=True, init=False)
+class TensorProductShape(Shape):
+    """
+    .. attribute:: bases
+
+        A :class:`tuple` of base shapes that form the tensor product.
+    """
+
+    def __init__(self, bases: Tuple[Shape, ...]) -> None:
```

Comment:
* Do we genuinely allow any shapes? Triangle x Triangle? That's already 4D, so I'm kind of short on imagination in terms of what that would look like/be useful for.
* Flatten if any of the underlying things is also a `TensorProduct`?

Reply:
> Do we genuinely allow any shapes? Triangle x Triangle? That's already 4D, so I'm kind of short on imagination in terms of what that would look like/be useful for.

There's a good chance a few places will break on some random list of shapes, but I'm not sure how to restrict the choices. Maybe just limit the resulting dimension to 3?

Reply:
How about "at most one non-line and otherwise only lines"?

Reply:
That sounds good! Implemented something like that in b7db9ff. I think the checks are ok, but definitely needs a closer look.